File: GoRide/src/main/java/com/pfeproject/GoRide/services/NotificationService.java
package com.pfeproject.GoRide.services;

import com.pfeproject.GoRide.entities.Notification;
import com.pfeproject.GoRide.entities.UserEntity;
import com.pfeproject.GoRide.repositories.NotificationRepository;
import com.pfeproject.GoRide.repositories.UserRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class NotificationService {

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private UserRepo userRepo;

    @Autowired
    @Lazy
    private EmailService emailService;

    @Transactional
    public Notification createNotification(Long userId, String title, String message, String type, String targetUrl) {
        UserEntity user = userRepo.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + userId));

        Notification notification = Notification.builder()
                .user(user)
                .title(title)
                .message(message)
                .type(type)
                .targetUrl(targetUrl)
                .isRead(false)
                .build();

        Notification saved = notificationRepository.save(notification);

        if (Boolean.TRUE.equals(user.getNotifEmail())
                && user.getEmail() != null && !user.getEmail().isBlank()) {
            emailService.sendNotificationEmail(
                    user.getEmail(),
                    user.getFirstName(),
                    title,
                    message,
                    targetUrl
            );
        }

        return saved;
    }

    public List<Notification> getNotificationsByUser(Long userId) {
        return getNotificationsByUser(userId, null);
    }

    public List<Notification> getNotificationsByUser(Long userId, String mode) {
        List<Notification> all = notificationRepository.findByUserIdOrderByCreatedAtDesc(userId);
        if (mode == null || mode.isBlank()) {
            return all;
        }
        return all.stream()
                .filter(n -> matchesClientMode(n, mode))
                .toList();
    }

    public long getUnreadCount(Long userId) {
        return countUnreadNotifications(notificationRepository.findByUserIdOrderByCreatedAtDesc(userId), null);
    }

    /**
     * Compte les notifications non lues visibles pour le mode client (TENANT = locataire, PASSENGER = passager).
     * Aligné sur le filtre de la page notifications Angular.
     */
    public long getUnreadCountForMode(Long userId, String mode) {
        return countUnreadNotifications(
                notificationRepository.findByUserIdOrderByCreatedAtDesc(userId),
                mode);
    }

    private long countUnreadNotifications(List<Notification> notifications, String mode) {
        return notifications.stream()
                .filter(n -> !Boolean.TRUE.equals(n.getIsRead()))
                .filter(n -> matchesClientMode(n, mode))
                .count();
    }

    private boolean matchesClientMode(Notification n, String mode) {
        if (mode == null || mode.isBlank()) {
            return true;
        }
        String text = ((n.getTitle() != null ? n.getTitle() : "") + " "
                + (n.getMessage() != null ? n.getMessage() : "")).toLowerCase();
        boolean hasPassengerKeyword = text.contains("course") || text.contains("chauffeur")
                || text.contains("trajet") || text.contains("passager") || text.contains("covoiturage")
                || text.contains("reservation") || text.contains("réservation")
                || text.contains("demande") || text.contains("inscription")
                || text.contains("partagé") || text.contains("partage");
        boolean hasRentalKeyword = text.contains("location") || text.contains("véhicule")
                || text.contains("vehicule") || text.contains("contrat")
                || text.contains("propriétaire") || text.contains("proprietaire")
                || text.contains("loué") || text.contains("loue") || text.contains("louer")
                || text.contains("caution") || text.contains("louée");

        if ("TENANT".equalsIgnoreCase(mode)) {
            return hasRentalKeyword || !hasPassengerKeyword;
        }
        if ("PASSENGER".equalsIgnoreCase(mode)) {
            return !hasRentalKeyword;
        }
        return true;
    }

    @Transactional
    public void markAsRead(Long notificationId) {
        notificationRepository.findById(notificationId).ifPresent(n -> {
            n.setIsRead(true);
            notificationRepository.save(n);
        });
    }

    @Transactional
    public void markAllAsRead(Long userId) {
        List<Notification> unread = notificationRepository.findByUserIdOrderByCreatedAtDesc(userId)
                .stream()
                .filter(n -> !n.getIsRead())
                .toList();

        unread.forEach(n -> n.setIsRead(true));
        notificationRepository.saveAll(unread);
    }

    @Transactional
    public void deleteNotification(Long id) {
        notificationRepository.deleteById(id);
    }
}

File: angularpfe/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { AuthService } from '../auth/auth.service';
import { RoleService } from '../auth/role.service';

export interface AppNotification {
  id: number;
  type: string;
  label: string;
  description: string;
  time: string;
  icon: string;
  color: string;
  isRead: boolean;
  targetUrl?: string;
}

@Injectable({
  providedIn: 'root'
})
export class NotificationService {
  private notificationsSubject = new BehaviorSubject<AppNotification[]>([]);
  public notifications$ = this.notificationsSubject.asObservable();
  
  private unreadCountSubject = new BehaviorSubject<number>(0);
  public unreadCount$ = this.unreadCountSubject.asObservable();

  constructor(
    private http: HttpClient,
    private authService: AuthService,
    private roleService: RoleService
  ) {
    if (this.authService.isLoggedIn()) {
      this.loadAll();
    }
    this.roleService.activeRole$.subscribe(() => {
      if (this.authService.isLoggedIn()) {
        this.loadAll();
      }
    });
  }

  /** @deprecated Préférer loadAll() pour garder badge et liste synchronisés. */
  setUnreadCount(_count: number): void {
    this.loadAll();
  }

  private notificationModeParam(): string | undefined {
    const role = this.roleService.getActiveRole();
    if (role === 'ROLE_CLIENT') return 'TENANT';
    if (role === 'ROLE_USER') return 'PASSENGER';
    return undefined;
  }

  public loadAll(): void {
    if (!this.authService.isLoggedIn()) return;

    this.getMyNotifications().subscribe({
      next: (notifs) => {
        const mapped = (notifs || []).map(n => this.mapNotification(n));
        // Le backend filtre déjà par ?mode= ; éviter un second filtre qui masque des notifs.
        const mode = this.notificationModeParam();
        const filtered = mode ? mapped : this.filterForActiveRole(mapped);
        this.notificationsSubject.next(filtered);
        this.unreadCountSubject.next(filtered.filter(n => !n.isRead).length);
      },
      error: (err) => console.error('Error loading notifications:', err)
    });
  }

  /** Aligné sur le filtre backend / page notifications (TENANT vs PASSENGER). */
  filterForActiveRole(notifs: AppNotification[]): AppNotification[] {
    const mode = this.notificationModeParam();
    if (!mode) return notifs;
    return notifs.filter(n => this.matchesClientMode(n, mode));
  }

  private matchesClientMode(n: AppNotification, mode: string): boolean {
    const text = `${n.label || ''} ${n.description || ''}`.toLowerCase();
    const hasPassengerKeyword =
      text.includes('course') ||
      text.includes('chauffeur') ||
      text.includes('trajet') ||
      text.includes('passager') ||
      text.includes('covoiturage') ||
      text.includes('réservation') ||
      text.includes('reservation');
    const hasRentalKeyword =
      text.includes('location') ||
      text.includes('véhicule') ||
      text.includes('vehicule') ||
      text.includes('contrat') ||
      text.includes('propriétaire') ||
      text.includes('proprietaire') ||
      text.includes('loué') ||
      text.includes('loue') ||
      text.includes('caution');

    if (mode === 'TENANT') {
      return hasRentalKeyword || !hasPassengerKeyword;
    }
    if (mode === 'PASSENGER') {
      return !hasRentalKeyword;
    }
    return true;
  }

  private mapNotification(backendNotif: any): AppNotification {
    let color = 'blue';
    let icon = 'ion-md-information-circle';
    
    const type = backendNotif.type ? backendNotif.type.toUpperCase() : 'INFO';
    if (type === 'SUCCESS') {
      color = 'green';
      icon = 'ion-md-checkmark-circle';
    } else if (type === 'WARNING') {
      color = 'orange';
      icon = 'ion-md-warning';
    } else if (type === 'DANGER') {
      color = 'red';
      icon = 'ion-md-close-circle';
    } else if (type === 'INFO') {
      color = 'blue';
      icon = 'ion-md-information-circle';
    } else {
      color = 'blue';
      icon = 'ion-md-notifications';
    }

    return {
      id: backendNotif.id,
      type: backendNotif.type || 'info',
      label: backendNotif.title || '',
      description: backendNotif.message || '',
      time: backendNotif.createdAt || new Date().toISOString(),
      icon: icon,
      color: color,
      isRead: !!(backendNotif.isRead ?? backendNotif.read),
      targetUrl: backendNotif.targetUrl
    };
  }

  getMyNotifications(): Observable<any[]> {
    const mode = this.notificationModeParam();
    const url = mode
      ? `${environment.apiUrl}/notifications/me?mode=${mode}`
      : `${environment.apiUrl}/notifications/me`;
    return this.http.get<any[]>(url);
  }

  getUnreadCount(): Observable<{ count: number }> {
    const mode = this.notificationModeParam();
    const url = mode
      ? `${environment.apiUrl}/notifications/me/unread-count?mode=${mode}`
      : `${environment.apiUrl}/notifications/me/unread-count`;
    return this.http.get<{ count: number }>(url);
  }

  markAsRead(id: number): void {
    this.http.put(`${environment.apiUrl}/notifications/${id}/read`, {}).subscribe({
      next: () => {
        this.loadAll();
      },
      error: (err) => console.error('Error marking notification as read:', err)
    });
  }

  markAllAsRead(): void {
    this.http.put(`${environment.apiUrl}/notifications/read-all`, {}).subscribe({
      next: () => {
        this.loadAll();
      },
      error: (err) => console.error('Error marking all notifications as read:', err)
    });
  }

  deleteNotification(id: number): void {
    this.http.delete(`${environment.apiUrl}/notifications/${id}`).subscribe({
      next: () => {
        this.loadAll();
      },
      error: (err) => console.error('Error deleting notification:', err)
    });
  }
}
